[CS-4829][CS-4830] Update reward center to list all claims and allow claiming one-by-one

oidoug opened the pull request:
### Description

This PR adds a new API resource to fetch individual reward proofs for given program and it's info from the SDK and refactor Reward Center Screen to display the program explainer and the individual proofs.

Notes:
Total claim available item is not being shown because it was too confusing, it's hard to indicate it's the total of the below cells, so I prefer to keep only the individual proofs visible.

The UI is WIP, I appreciate feedbacks. It's still not showing the proof explanation, coming in the next PR.

- [x] Completes #(CS-4829)
- [x] Completes #(CS-4830)

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS
- [x] Tested on Android

### Screenshots

<img width="300" alt="Screenshot" src="https://user-images.githubusercontent.com/129619/202204171-1c051565-e512-4a63-b9b1-11c01053a570.png">

<img width="300" alt="Screenshot" src="https://user-images.githubusercontent.com/129619/202204259-e7db20eb-2c14-437c-b8ab-b2a2d6ddaf59.png">

oidoug committed:
Adds API resource to fetch individual reward proofs for given program;
Adds API resource to fetch program information;
Refactor Reward Center Screen to display program info and individual proofs;

oidoug committed:
Removes debug logs

oidoug committed:
Renames text variant to be closer to porpuse;

oidoug committed:
Improves loading state for reward program banner

paulinhapenedo commented on cardstack/src/screens/RewardsCenterScreen/RewardsCenterScreen.tsx:
```diff
@@ -18,49 +16,57 @@ const RewardsCenterScreen = () => {
           <Container flex={1}>
             {!isRegistered &&
-              (hasRewards ? (
+              (mainPoolRowProps ? (
```
If we are taking the proposed approach, we should probably renamed this because `mainPoolRowProps` isn't conveying its purpose on a if statement.

oidoug replied:
Yeah, the naming used all around rewards is mislanding. 

Checking for the full "pool balance" may not even be needed anymore, but better keep it for now because we aren't sure yet what will change in rewards.

danibonilha replied:
I'm confused about the usage of mainPoolRewards here too, `hasRewards` in the render gives a clearer idea why we have the condition, but just checking for `mainPoolRowProps` doesn't say much, and we need to go into the prop to see that is checking if it hasRewards

paulinhapenedo commented:
I'm getting an error when trying to claim a reward that I have on Gnosis. The error is probably because it's dust as I already have a rewards safe.


https://user-images.githubusercontent.com/690904/202302369-fdda2b9e-1290-4a3d-a587-54264b76e2cd.mp4

danibonilha commented on cardstack/src/screens/RewardsCenterScreen/RewardsCenterScreen.tsx:
```diff
@@ -18,49 +16,57 @@ const RewardsCenterScreen = () => {
+        coinSymbol: reward?.token.symbol || '',
+        isClaimable: true,
+      })),
+    [rewards]
```
I think this is okay for now, but with the new UI we might need a new component, instead of reusing the rewardRow, in this case, i'd suggest that ClaimContent, gets the rewards and deals with the formatting within it

oidoug replied:
Seems fine to use RewardRow, I'm using it for now without changes. Using your suggestion of handling the reward list inside ClaimContent.

danibonilha commented on cardstack/src/screens/RewardsCenterScreen/components/ClaimContent.tsx:
```diff
@@ -42,7 +42,7 @@ export const ClaimContent = ({ claimList }: ClaimContentProps) => {
+  const ClaimList = useMemo(
     () =>
       claimList?.map((item, index) => (
         <RewardRow
```
Now that we can have a bunch of rows with information, this should be converted into a flatlist

danibonilha commented on cardstack/src/theme/textVariants.ts:
```diff
@@ -79,4 +79,11 @@ export const textVariants = {
+    textTransform: 'uppercase',
+    textAlign: 'center',
+    ...fontFamilyVariants.semiBold,
+  },
```
Is there an intention to reuse this ? If not, I don't think it makes much sense to add this kind of text variables

oidoug replied:
I wanted to reuse in the promo banner, but it ended up being slightly different.. removing.

danibonilha commented on cardstack/src/services/rewards-center/rewards-center-types.ts:
```diff
@@ -75,3 +80,28 @@ export interface RewardWithdrawParams
+  accountAddress: string;
+  rewardProgramId: string;
+  nativeCurrency: NativeCurrency;
+}
```
I guess you could reuse `RewardsSafeQueryParams` giving that safeAddress is optional and you are reusing the data.query anyways on the component

danibonilha commented on cardstack/src/services/rewards-center/rewards-center-service.ts:
```diff
@@ -250,6 +282,14 @@ export const getWithdrawGasEstimate = async ({
+
+  const programInfo = await rewardManager.getRewardProgramInfo(rewardProgramId);
+
+  return programInfo.programExplainer;
```
Maybe we should reflect that this function just returns the `programExplainer`, or we should return the entire `getRewardProgramInfo`  and use the programExplainer on the component

danibonilha commented on cardstack/src/services/rewards-center/rewards-center-types.ts:
```diff
@@ -75,3 +80,28 @@ export interface RewardWithdrawParams
+      display: string;
+    };
+  };
+};
```
There's a `TokenType` to extend and a `TokenInfo`

oidoug committed:
Adds new RewardClaimSingle screen and hook to handle single reward claim;
Refactors proof gas estimate and claim to take reward list;
Updates ClaimContent to use FlatList for reward items;
Updates Types to use SDK's ClaimableProof type (that includes safe and gas);
Updates RewardProofType to use TokenType;

oidoug committed:
Rebases with develop;
Fixes linting issues.

oidoug committed:
Removes unused mainPool prop.

oidoug committed:
Removes textVariant only used once.

oidoug committed:
Reusing type RewardsSafeQueryParams instead of creating new.

oidoug committed:
Renames getRewardProgramInfo to getRewardProgramExplaniner.

oidoug committed:
Fixes incorrect func component ListFooter call.

oidoug commented on cardstack/src/screens/RewardsCenterScreen/components/ClaimContent.tsx:
```diff
@@ -33,46 +29,36 @@ const tabs = [
+    [title]
+  );
+
+  const ListFooter = () => (
```
Note: useMemo here did not work, the linter errors with `TypeError: Cannot read property 'parent' of null`. It looks like FlatList doesn't work with a memoized footer, but I can't find the specific reason.

Extracting this as a component looks like too much.
I appreciate your input.

danibonilha replied:
Not sure why the linter would error this, but it might be because our typescript is 4.5.5 and it supports <4.5.0 only, giving this is the footer and it's small, maybe we could assign the <Container> and it's content directly to the ListFooterComponent, since the Flatlist component is small and the footer is the last parameter

oidoug committed:
Adds check for isClaimable for RewardRows.

danibonilha commented on cardstack/src/services/rewards-center/rewards-center-service.ts:
```diff
@@ -78,6 +79,55 @@ const getValidProofs = async ({
+}: RewardsSafeQueryParams): Promise<RewardValidProofsResult> => {
+  const rewardPoolInstance = await getRewardsPoolInstance();
+
+  const proofs = await rewardPoolInstance.getUnclaimedValidProofs(
+    accountAddress,
+    rewardProgramId
+  );
+
+  const claimableProofs = safeAddress
+    ? await rewardPoolInstance.getUnclaimedValidProofsWithoutDust(
+        accountAddress,
+        rewardProgramId,
+        safeAddress
+      )
+    : [];
+
```
isn't both of these proofs claimables?  Maybe we could rename to give better context

oidoug replied:
Kinda, `proofs` might contain dust that we can't claim. `claimableProofs`'s proofs should all be claimable since it doesn't include dust.

danibonilha commented on cardstack/src/services/rewards-center/rewards-center-service.ts:
```diff
@@ -78,6 +79,55 @@ const getValidProofs = async ({
+      const claimingInfo =
+        claimableProofs.find(
+          findProof => findProof.rootHash === proof.rootHash
+        ) || {};
```
What does this mean ? if it exists on withoutDust it's claimable ? Maybe we could add a comment to refer to that

oidoug replied:
Thats was my reasoning, proofs without dust are claimable. I'll improve naming and add a comment 👍

danibonilha commented on cardstack/src/services/rewards-center/rewards-center-service.ts:
```diff
@@ -78,6 +79,55 @@ const getValidProofs = async ({
+        ...proof,
+        ...tokenWithPrice,
+        ...claimingInfo,
+        isClaimable: !!claimingInfo,
```
I'm confused on spreading proof and claimingInfo, why do we need that ?

oidoug replied:
`claimingInfo` has all values `proof` has plus some extra values, like estimatedGas and `tokenSymbol`. It could replace proof, but spreading and replacing looks cleaner IMO.

danibonilha commented on cardstack/src/screens/RewardsCenterScreen/components/ClaimContent.tsx:
```diff
@@ -81,6 +67,33 @@ export const ClaimContent = ({ claimList }: ClaimContentProps) => {
+
+  return (
+    <FlatList
+      data={rewards}
```
Let's add `showsVerticalScrollIndicator={false}` to avoid reintroducing the scroll bug

oidoug replied:
The scroll bug happens because the parent container had a padding. No the padding happens inside each FlatList's children and the scroll bar is presented in the correct place, so I preferred to show it.

danibonilha reviewed (approved):
Just added some comments to be addressed before merging it

oidoug committed:
Better naming and comments for fetchValidProofsWithToken service call.

oidoug committed:
Refactors Footer for FlatList to be inplace
